### PR TITLE
Let RenderEditable.delete directly delete from the controller text.

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
Fixes `delete` and `deleteForward` so they operate directly on `TextEditingController.value`. Internal: b/187337775

Related: https://github.com/flutter/flutter/issues/80226

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/rendering/editable.dart [hunk 1]
```diff
@@ -1011,12 +1011,21 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
+  //
+  // Operates on the text/selection contained in textSelectionDelegate, and does
+  // not depend on `RenderEditable.selection`.
+  //
+  // If the selection is currently non-empty, this method deletes the selected
+  // text and returns true. Otherwise this method does nothing and returns
+  // false.
+  bool _deleteNonEmptySelection(SelectionChangedCause cause) {
+    assert(!readOnly);
+    final TextSelection selection = textSelectionDelegate.textEditingValue.selection;
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/editable.dart:1043]: We have used the `textSelectionDelegate.textEditingValue` in a lot of places.
maybe we should define a getter for textEditingValue?

File: packages/flutter/lib/src/rendering/editable.dart [hunk 2]
```diff
@@ -1091,10 +1101,10 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
 
   /// Deletes backwards from the current selection.
   ///
-  /// If the [selection] is collapsed, deletes a single character before the
+  /// If the selection is collapsed, deletes a single character before the
   /// cursor.
   ///
-  /// If the [selection] is not collapsed, deletes the selection.
+  /// If the selection is not collapsed, deletes the selection.
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/editable.dart:1126]:
```suggestion
  /// If the selection in the [textSelectionDelegate.textEditingValue] is not collapsed, deletes the selection.
```

File: packages/flutter/lib/src/rendering/editable.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:1043]: I'm thinking we may want to get rid of it in `RenderEditable`: https://github.com/flutter/flutter/issues/79420#issuecomment-837456362

File: packages/flutter/lib/src/rendering/editable.dart [hunk 3]
```diff
@@ -1091,10 +1101,10 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
 
   /// Deletes backwards from the current selection.
   ///
-  /// If the [selection] is collapsed, deletes a single character before the
+  /// If the selection is collapsed, deletes a single character before the
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:1137]: I added a section before this to describe which "selection" this operates on. The documentation is still lacking but I think we want to get rid of these 2 methods from `RenderEditable` anyways.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 4]
```diff
@@ -1240,40 +1275,39 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
+  /// If the selection is not collapsed, deletes the selection.
   ///
   /// {@macro flutter.rendering.RenderEditable.cause}
   ///
   /// See also:
   ///
   ///   * [delete], which is same but in the opposite direction.
   void deleteForward(SelectionChangedCause cause) {
-    assert(_selection != null);
+    // TODO(LongCatIsLooong): remove this method from RenderEditable.
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:1319]: I tried moving these 2 methods to default text editing actions. It works I think, but`TextEditingActionTarget` only exposes the `RenderEditable` so it feels a bit awkward. I think we may want to add more things to `TextEditingActionTarget`, or just use the current input client (not sure if possible).

File: packages/flutter/lib/src/rendering/editable.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/editable.dart:1138]: I see, my first comment still apply I think? or did i miss something?

[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:1138]: Sorry missed that part. If the selection is collapsed the expected behavior is to delete the character immediately before the caret. I think there's a test for this?

**Review by chunhtai — APPROVED** [review]
LGTM

**Pull request merged**